src/controller/java/src/matter/controller/MatterControllerImpl.kt

```diff
@@ -0,0 +1,661 @@
+        devicePtr,
+        chipAttributePaths,
+        chipEventPaths,
+        false,
+        CHIP_IM_TIMEOUT_MS,
+      )
+    }
+  }
+
+  override suspend fun write(writeRequests: WriteRequests): WriteResponse {
+    // To prevent potential issues related to concurrent modification, assign
+    // the value of the mutable property 'nodeId' to a temporary variable.
+    val nodeId = this.nodeId
+    check(nodeId != null) { "nodeId has not been initialized yet" }
+
+    val devicePtr: Long = getConnectedDevicePointer(nodeId)
+
+    val attributeWriteRequests =
+      writeRequests.requests.map { request ->
+        AttributeWriteRequest.newInstance(
+          ChipPathId.forId(request.attributePath.endpointId.toLong()),
+          ChipPathId.forId(request.attributePath.clusterId.toLong()),
+          ChipPathId.forId(request.attributePath.attributeId.toLong()),
+          request.tlvPayload
+        )
+      }
+
+    val failures = mutableListOf<AttributeWriteError>()
+
+    return suspendCancellableCoroutine { continuation ->
+      val writeCallback =
+        object : WriteAttributesCallback {
+          override fun onResponse(attributePath: ChipAttributePath) {
+            logger.log(Level.INFO, "write success for attributePath:%s", attributePath.toString())
+          }
+
+          override fun onError(attributePath: ChipAttributePath?, ex: Exception) {
+            logger.log(
+              Level.SEVERE,
+              "Failed to write attribute at path: %s",
+              attributePath.toString()
+            )
+
+            if (attributePath == null) {
+              if (ex is ChipDeviceControllerException) {
+                continuation.resumeWithException(
+                  MatterControllerException(ex.errorCode, ex.message)
+                )
+              } else {
+                continuation.resumeWithException(ex)
+              }
+            } else {
+              failures.add(AttributeWriteError(attributePath.wrap(), ex))
+            }
+          }
+
+          override fun onDone() {
+            logger.log(Level.INFO, "writeAttributes onDone is received")
+
+            if (failures.isNotEmpty()) {
+              continuation.resume(WriteResponse.PartialWriteFailure(failures))
+            } else {
+              continuation.resume(WriteResponse.Success)
+            }
+          }
+        }
+
+      deviceController.write(
+        writeCallback,
+        devicePtr,
+        attributeWriteRequests.toList(),
+        writeRequests.timedRequest?.toMillis()?.toInt() ?: 0,
+        CHIP_IM_TIMEOUT_MS,
+      )
+    }
+  }
+
+  override suspend fun invoke(request: InvokeRequest): InvokeResponse {
+    // To prevent potential issues related to concurrent modification, assign
+    // the value of the mutable property 'nodeId' to a temporary variable.
+    val nodeId = this.nodeId
+    check(nodeId != null) { "nodeId has not been initialized yet" }
+
+    val devicePtr: Long = getConnectedDevicePointer(nodeId)
```

Comment:
we do not seem to need a local copy of nodeId

Reply:
same as above